src/lua/uri.c

```diff
@@ -0,0 +1,313 @@
+							if_not_exist);
+		} else {
+			rc = -1;
+		}
+		lua_pop(L, 1);
+	}
+	lua_pop(L, 1);
+	return rc;
+}
+
+/**
+ * Check if there is a field with the name @name in the table,
+ * which located at the top of the lua stack.
+ */
+static bool
+is_field_present(struct lua_State *L, const char *name)
+{
+	lua_pushstring(L, name);
+	lua_rawget(L, -2);
+	bool field_is_present = (lua_type(L, -1) != LUA_TNIL);
+	lua_pop(L, 1);
+	return field_is_present;
+}
+
+/**
+ * Check if there are nested tables in the table, which located
+ * at the top of the lua stack.
+ */
+static bool
+is_nested_uri_tables_present(struct lua_State *L)
+{
+	bool nested_tables_is_present = false;
+	int uri_count = lua_objlen(L, -1);
+	for (int i = 0; i < uri_count && !nested_tables_is_present; i++) {
+		lua_rawgeti(L, -1, i + 1);
+		nested_tables_is_present = (lua_type(L, -1) == LUA_TTABLE);
+		lua_pop(L, 1);
+	}
+	return nested_tables_is_present;
+}
+
+/**
+ * Perform some checks for the URI table. The types of checks depend on
+ * @a table_is_nested.
+ */
+static bool
+uri_table_is_valid(struct lua_State *L, bool table_is_nested)
+{
+	/** All numerical keys in URI table corresponds to URI */
+	int uri_count_from_numerical_keys = lua_objlen(L, -1);
+	bool params_is_present = is_field_present(L, "params");
+	bool uri_key_is_present = is_field_present(L, "uri");
+	bool nested_uri_tables_is_present = is_nested_uri_tables_present(L);
+	bool default_params_is_present = is_field_present(L, "default_params");
+
+        /* Should be at least one URI in URI table */
+	if (uri_count_from_numerical_keys == 0 && !uri_key_is_present)
+		return false;
+	/* Mix "uri=" syntax and URIs from numerical keys is banned */
+	if (uri_count_from_numerical_keys != 0 && uri_key_is_present)
+		return false;
+	/* Should be no more than one URI in nested URI table */
+	if (uri_count_from_numerical_keys > 1 && table_is_nested)
+		return false;
+	/* Table "params" is now allowed for several URIs */
+	if (uri_count_from_numerical_keys > 1 && params_is_present)
+		return false;
+	/* Nested tables is not allowed in nested tables */
+	if (table_is_nested && nested_uri_tables_is_present)
+		return false;
+	/* Table "default_params" is not allowed in nested tables */
+	if (default_params_is_present && table_is_nested)
+		return false;
+	/*
+	 * Table "params" is not allowed in table, which contains,
+	 * nested URI tables.
+	 */
+	if (params_is_present && nested_uri_tables_is_present)
+		return false;
+	return true;
+}
+
+/**
+ * Create @a uri_set from the table, which located at the top of lua stack.
+ * Supports many different formats.
+ */
+static int
+uri_set_create_from_lua(struct uri_set *uri_set, struct lua_State *L,
+			bool table_is_nested)
+{
+	int rc = 0;
+	struct uri_set u = { 0 };
+
+	/** Some preliminary checks of the URI table. */
+	if (!uri_table_is_valid(L, table_is_nested))
+		return -1;
+
+	lua_pushnil(L);
+	while (lua_next(L, lua_gettop(L) - 1) != 0 && rc == 0) {
```

Comment:
AFAIU this is an array so let's please use `lua_rawgeti` and `lua_rawlen` for iterating over uris stored in a Lua table. You can use `lua_rawget` for retrieving additional query parameters. This should simplify this loop.

Reply:
fixed